File: src/nbla/function/generic/embed.cpp
// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/** Embed
 */
#include <nbla/array.hpp>
#include <nbla/function/embed.hpp>
#include <nbla/variable.hpp>

#include <algorithm>
#include <cstring>

namespace nbla {

NBLA_REGISTER_FUNCTION_SOURCE(Embed);

template <typename T, typename T1>
void Embed<T, T1>::setup_impl(const Variables &inputs,
                              const Variables &outputs) {
  Shape_t shape_x = inputs[0]->shape();
  Shape_t shape_w = inputs[1]->shape();
  Shape_t shape_y = shape_x;
  shape_y.insert(shape_y.end(), shape_w.begin() + 1, shape_w.end());
  outputs[0]->reshape(shape_y, true);
}

template <typename T, typename T1>
void Embed<T, T1>::forward_impl(const Variables &inputs,
                                const Variables &outputs) {
  const T *x = inputs[0]->get_data_pointer<T>(this->ctx_);
  const T1 *w = inputs[1]->get_data_pointer<T1>(this->ctx_);
  T1 *y = outputs[0]->cast_data_and_get_pointer<T1>(this->ctx_, true);

  Size_t stride0 = inputs[1]->size(1);
  for (int i = 0; i < inputs[0]->size(); ++i) {
    memcpy((void *)(y + i * stride0), w + x[i] * stride0, sizeof(T1) * stride0);
  }
}

template <typename T, typename T1>
void embed_backward_cpu(int size, int stride0, T1 *dw, const T1 *dy,
                        const T *x) {
  for (int i = 0; i < size; ++i) {
    for (int j = 0; j < stride0; ++j) {
      dw[x[i] * stride0 + j] += dy[i * stride0 + j];
    }
  }
}

template <typename T, typename T1>
void Embed<T, T1>::backward_impl(const Variables &inputs,
                                 const Variables &outputs,
                                 const vector<bool> &propagate_down,
                                 const vector<bool> &accum) {

  NBLA_CHECK(!propagate_down[0], error_code::value,
             "Index array can not be propagated down.");
  if (!propagate_down[1]) {
    return;
  }
  if (!accum[1])
    inputs[1]->grad()->zero();
  const T *x = inputs[0]->get_data_pointer<T>(this->ctx_);
  T1 *dw = inputs[1]->cast_grad_and_get_pointer<T1>(this->ctx_, false);
  const T1 *dy = outputs[0]->get_grad_pointer<T1>(this->ctx_);
  Size_t stride0 = inputs[1]->size(1);
  embed_backward_cpu(inputs[0]->size(), stride0, dw, dy, x);
}

} // namespace nbla

File: python/src/nnabla/backward_function/embed.py
# Copyright 2019,2020,2021 Sony Corporation.
# Copyright 2021 Sony Group Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import nnabla as nn
import nnabla.function as _F
import nnabla.functions as F

# from nnabla.function import PythonFunction
from .backward_function import LinearFilterGrad


class EmbedFilterGrad(LinearFilterGrad):

    def __init__(self, ctx, base_axis=1):
        super(EmbedFilterGrad, self).__init__(ctx)
        self._linear = _F.Embed(ctx)

    def backward_impl(self, inputs, outputs, propagate_down=[], accum=[]):
        if not propagate_down[0]:
            return

        dy = inputs[0].data
        x0 = inputs[1].data
        dw0 = outputs[0].data

        gdy = inputs[0].grad
        gx0 = inputs[1].grad
        gdw0 = outputs[0].grad

        inputs_fwd, outputs_fwd = self._create_fwd_inputs_outputs(
            inputs, outputs)
        vx = inputs_fwd[0].apply(need_grad=False)
        vw = inputs_fwd[1].apply(need_grad=False)
        vy = outputs_fwd[0]

        # w.r.t. x0
        # do nothing since x0 is the index.

        # w.r.t. dy
        vx.data = x0
        vw.data = gdw0
        if accum[0]:
            self._linear.forward(inputs_fwd, outputs_fwd)
            gdy += vy.data
        else:
            vy.data = gdy
            self._linear.forward(inputs_fwd, outputs_fwd)


def embed_backward(grad_inputs, inputs, input_shapes, outputs, output_shapes):
    """
    Args:
      grad_inputs (list of :obj:`nnabla.Variable`): Propagated grads to this backward function.
      inputs (list of :obj:`nnabla.Variable` and None): Input Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      input_shapes (list of tuple of :obj:`int`): Input shapes of the forward function.
          The shapes of the inputs in which None is set can be passed.
      outputs (list of :obj:`nnabla.Variable` and None): Output Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      output_shapes (list of tuple of :obj:`int`): Output shapes of the forward function.
          The shapes of the outputs in which None is set can be passed.
      kwargs (dict of arguments): Dictionary of the corresponding function arguments.

    Return:
      list of Variable: Return the gradients wrt inputs of the corresponding function.
    """
    dy = grad_inputs[0]
    x0 = inputs[0]
    w0_shape = input_shapes[1]

    ctx = nn.get_current_context()
    dfw = EmbedFilterGrad(ctx)
    dfw.wshape = w0_shape

    dw0 = dfw(dy, x0)
    return None, dw0


def embed_filter_grad_backward(grad_inputs, inputs, input_shapes, outputs, output_shapes):
    """
    Args:
      grad_inputs (list of :obj:`nnabla.Variable`): Propagated grads to this backward function.
      inputs (list of :obj:`nnabla.Variable` and None): Input Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      input_shapes (list of tuple of :obj:`int`): Input shapes of the forward function.
          The shapes of the inputs in which None is set can be passed.
      outputs (list of :obj:`nnabla.Variable` and None): Output Variables of the forward function
          if this backward function depends on it. Otherwise, None is set instead.
      output_shapes (list of tuple of :obj:`int`): Output shapes of the forward function.
          The shapes of the outputs in which None is set can be passed.
      kwargs (dict of arguments): Dictionary of the corresponding function arguments.

    Return:
      list of Variable: Return the gradients wrt inputs of the corresponding function.
    """
    gdw = grad_inputs[0]
    dy = inputs[0]
    x0 = inputs[1]
    gdy = F.embed(x0, gdw)
    return gdy, None
